### DISCUSSION TITLE
Project Organization - Multiple Repos

### DISCUSSION THREAD

**Discussion by Brennan1994** [HydrologicEngineeringCenter/HEC-FDA#275] [Q&A] [answered]
Initially we split out our project into separate repos. currently, we have 
- HEC-FDA (Official UI)
- fda-model 
- fda-statistics
- fda-utilities

In splitting these out, I think it helped isolate problems. It let us work separately from Cody for awhile, and make progress when we needed to. Now, I think it may be more a burden than it's worth. There's a couple examples where this organization makes our lives more painful than necessary

1. Updating a lower level library with api breaking changes: Removing something from statistics, breaks model, which breaks view model, which breaks view. It's not immediately obvious whats going to go wrong, or where to fix it.
2. Updating a lower level library takes a long time to implement into a higher level. For a statistics update, you have to pull request on statistics, finish checks, merge, finish checks to publish, then update model, pull request, finish checks, merge, finish checks to publish, then update UI. That's a lot of steps. 

I'm wondering if it would be worth it to consolidate some of these models. It seems reasonable to me that all both model and statistics could live within the FDA repo again now that so much of the old junk has been cleaned out. I'm in no hurry to make this happen, but wanted to start the conversation after some pain Cody and I had making updates yesterday.

**Comment by HenryGeorgist** [comment]
I think we could collapse fda-statistics into fda-model with no significant repercussions.

I'd like to get rid of fda-utilities completely. I think the biggest issue there is the messaging system which is only partly used in HEC-FDA

I think if we do those two, then we could consider moving fda-model into HEC-FDA, or keeping it separate after some time working with it that way and considering the pros and cons.

**Comment by rnugent3** [comment]
I like the idea of pursuing this incrementally and assessing as we go along. 

I would personally like to see fda-model and fda-statistics in the same repository. I ran into point 2 several times - e.g. working on model and I find a bug in statistics or vice-versa.

**Reply by HenryGeorgist** [reply]
the only concern I would have is the wiki on statistics - we don't want to orphan that content, it might be easy to just copy and paste. I'm not too concerned about the issues and tickets or git histories at this point. I think the statistics library is pretty "stable" at this point.

**Reply by rnugent3** [reply]
good point. the wiki is a few pages we could easily copy and paste, and we could transfer the issues to fda-model before making the move.

**Reply by HenryGeorgist** [reply]
transferring issues is a great idea! We could have pretty much anyone to do that.

**Comment by Brennan1994** [comment]
I concur with the staged approach. I think combining Statistics and Model into the same repo and solution would be pretty easy without harming anyone. Dropping utilities would also be wonderful. I'll volunteer to do the merger, since there will be some minor adjustments needed on the build script. I think we keep the fda-model repo, and trash the statistics one.  

I won't start the process until all the issues and Wiki have been copied out.

**Comment by Brennan1994** [comment]
fda-model and fda-statistics now exist in the fda-model repository as a result of this discussion. All issues and Wiki were migrated.

**Comment by Brennan1994** [comment]
Reviving this conversation after working on some ViewModel/Model Integrations. I'm finding it burdensome to keep these separated by Repo. While I'm developing in both simultaneously, I end up temporarily adding model to the FDA solution so that I can make changes and test as I go, then I have to strategically commit only certain files and revert others before I make a pull request. This is a super messy workflow.

I'd much rather just have these in the same solution all the time. I propose we consolidate, and move all the model libraries into the HEC-FDA repository and solution. 

Pros
- Easier debugging interface between model and viewmodel.
- One consolodated place for all FDA discussions and issues. 

Cons
- Migrating current issues and discussions over. 

I'm not sure I see any benefits anymore to our system now. I'm happy to be corrected if you see otherwise. I do not think this change has to be immediate.

**Reply by rnugent3** [reply]
I would like to see all of our projects in the same solution, too.

**Reply by codymccoy** [reply]
I would also like to see the model in the main fda solution. That would make it much easier to change things in the model and see those changes in the UI.

**Comment by Brennan1994** [comment, marked as answer]
The model projects, including fda-model, Statistics, and their test projects have been moved to the HEC-FDA repository
